File: atelier/src/lib/eidos/coffre-horaire.ts
/**
 * Coffre horaire — un coffre par bloc, une pièce par claim (docs/SPEC_COFFRE_HORAIRE.md).
 * Port à l'octet de labo/coffre_horaire.py ; la parité tient par la famille `coffre`
 * de vecteurs.json (vecteurs.test.ts).
 *
 *   graine = SHA-256d("eidos-coffre/1" ‖ id_bloc(32) ‖ txid(32) ‖ rang(4))
 *   tier   = 1 + zéros de tête du premier octet ∈ 1..9   (1/2, 1/4 … 1/256, 1/256)
 *   objets = tier OBJETS RÉELS : h_j = SHA-256d(graine ‖ j(1)), puis objetDepuisGraine + habille,
 *            exactement comme un tirage de bloc (inventaire.ts). Âge par tier.
 *
 * L'horloge est la tête signée, jamais l'horloge de la machine : un coffre existe parce
 * qu'un bloc existe. Une pièce ne réclame qu'un coffre par bloc (`tour.coffres`), et le
 * prend ENTIER : les t objets entrent dans `coffre.objets`, rien n'est perdu, aucun sac ne
 * borne un claim (décision A15, docs/FEUILLE_DE_ROUTE.md §3). L'ouverture rendue au joueur
 * (`Ouverture`) porte exactement ce que le juge a accepté — ni plus, ni moins.
 *
 * Figures, pas preuves : rien ici n'engage le carnet. Le tirage seul vit ici ; le juge
 * (`jugerClaim`) n'invente rien — la tête se vérifie par `temoin.ts` (XMSS) et la pièce par
 * `merkle.ts` (contre `utxoRoot`), exactement comme une ascension (`ancrage.ts`).
 * LIMITE : une pièce neuve peut réclamer tous les blocs passés qu'elle sait prouver ; rien ne
 * le borne hors « une pièce, un bloc ». La page n'offre que la tête suivie, et c'est une
 * politique d'interface, pas une règle. Voir spec §5.
 */

import { concat, fromHex, hexOf, sha256d, u32, utf8 } from "./hash.ts";
import { feuilleSortie, verifierPreuve, type PreuvePortable, type SortieMin } from "./merkle.ts";
import { objetDepuisGraine } from "./objets.ts";
import { habille } from "./equipement.ts";
import { SIGNATURES, type Signature } from "./signatures.ts";
import { verifierTeteReseau, type TeteReseau } from "./temoin.ts";
import type { Coffre, NomAge, ObjetPorte } from "./types.ts";

export const SPEC_COFFRE = "eidos-coffre/1";
export const TAG_COFFRE = utf8(SPEC_COFFRE);
export const TIERS = 9;

export const AGES_COFFRE = [
  "Kali",
  "Kali",
  "Kali",
  "Dvapara",
  "Dvapara",
  "Treta",
  "Treta",
  "Satya",
  "Satya",
] as const satisfies readonly NomAge[];

/** P(tier = t) : 2^−t pour t = 1..8, 2^−8 pour t = 9. Somme exactement 1. */
export const PROBA_TIER: readonly number[] = [
  ...Array.from({ length: 8 }, (_, i) => 2 ** -(i + 1)),
  2 ** -8,
];

export type CoffreTire = { graine: string; tier: number; objets: ObjetPorte[] };

/**
 * La muse d'un tier — une lecture (spec §3) : neuf tiers pour neuf muses, du plus commun
 * (1, Thalie, ⊕) au plus rare (9, Uranie, ★). SIGNATURES va d'Uranie (0) à Thalie (8),
 * donc le tier t lit SIGNATURES[9 − t]. Rien de plus : le tier ne change ni le mot ni les axes.
 */
export function museDuTier(tier: number): Signature {
  if (!Number.isInteger(tier) || tier < 1 || tier > TIERS)
    throw new Error(`tier ${tier} hors 1..${TIERS}`);
  return SIGNATURES[TIERS - tier]!;
}

function u8(n: number): Uint8Array {
  return new Uint8Array([n & 255]);
}

export function graineCoffre(
  idBlocHex: string,
  piece: Pick<SortieMin, "txid" | "rang">,
): Uint8Array {
  return sha256d(
    concat(TAG_COFFRE, fromHex(idBlocHex), fromHex(piece.txid), u32(piece.rang >>> 0)),
  );
}

/** Zéros de tête du premier octet, +1 : 255 → 1, 0 → 9. */
export function tierDe(graine: Uint8Array): number {
  const b = graine[0]!;
  return 1 + (b === 0 ? 8 : 8 - (32 - Math.clz32(b)));
}

export function coffreDe(
  idBlocHex: string,
  piece: Pick<SortieMin, "txid" | "rang">,
  hauteur = 0,
): CoffreTire {
  const g = graineCoffre(idBlocHex, piece);
  const tier = tierDe(g);
  const age = AGES_COFFRE[tier - 1]!;
  const objets: ObjetPorte[] = [];
  for (let j = 0; j < tier; j++) {
    const h = sha256d(concat(g, u8(j)));
    const o = objetDepuisGraine(h, age);
    objets.push(
      habille(
        {
          mot: o.mot,
          archetype: o.archetype,
          age: o.age,
          nonce: ((h[8]! << 8) | h[9]!) & 65535,
          hauteur,
        },
        h[10]!,
      ),
    );
  }
  return { graine: hexOf(g), tier, objets };
}

/** La clé d'un claim : une pièce, un bloc. */
export function cleClaim(idBlocHex: string, piece: Pick<SortieMin, "txid" | "rang">): string {
  return `${piece.txid}:${piece.rang}@${idBlocHex}`;
}

// ---------------------------------------------------------------------------
// Le juge : ce qui doit être vrai pour qu'un coffre soit réclamé.
// Rien de neuf — la tête par temoin.ts (XMSS), la pièce par merkle.ts, comme une ascension.

export type Claim = { tete: TeteReseau; piece: SortieMin; preuve: PreuvePortable };
export type VerdictClaim = { ok: true; idBloc: string; cle: string } | { ok: false; motif: string };

/**
 * Juge un claim, dans l'ordre où ça coûte le moins cher : déjà réclamé, puis la pièce contre la
 * racine, puis la signature de la tête. Le juge NE DIT PAS que la pièce est au joueur — cela se
 * prouve en la dépensant, comme pour une ascension (spec §6).
 */
export function jugerClaim(
  claim: Claim,
  federation: Parameters<typeof verifierTeteReseau>[1],
  deja: ReadonlySet<string>,
): VerdictClaim {
  const { tete, piece, preuve } = claim;
  const cle = cleClaim(tete.idBloc, piece);
  if (deja.has(cle)) return { ok: false, motif: "coffre déjà réclamé pour cette pièce à ce bloc" };
  if (preuve.racine !== tete.utxoRoot)
    return { ok: false, motif: "la preuve ne porte pas sur la racine de cette tête" };
  if (preuve.feuille !== hexOf(feuilleSortie(piece)))
    return { ok: false, motif: "la feuille n'est pas celle de cette pièce" };
  if (!verifierPreuve(preuve)) return { ok: false, motif: "chemin Merkle invalide" };
  const v = verifierTeteReseau(tete, federation);
  if (!v.ok) return { ok: false, motif: `tête refusée (${v.motif})` };
  return { ok: true, idBloc: tete.idBloc, cle };
}

/**
 * L'ouverture : ce qu'un claim accepté rend au joueur, et rien d'autre. Jamais persistée —
 * elle vit le temps d'être montrée ; les objets, eux, sont déjà dans `coffre.objets`.
 */
export type Ouverture = {
  cle: string;
  idBloc: string;
  hauteur: number;
  graine: string;
  tier: number;
  objets: ObjetPorte[];
};

export type Reclamation =
  | { ok: true; coffre: Coffre; ouverture: Ouverture }
  | { ok: false; motif: string };

/** Deux objets portés sont le même s'ils ont même mot, même hauteur et même teinte. */
export function memeObjet(a: ObjetPorte, b: ObjetPorte): boolean {
  return a.mot === b.mot && a.hauteur === b.hauteur && a.nonce === b.nonce;
}

/**
 * Réclamer pour de vrai : juge le claim, puis pose les t objets dans `coffre.objets` et la clé
 * dans `coffre.tour.coffres`. Le coffre est pris entier ; il n'est pas modifié en cas de refus.
 */
export function reclamerDansCoffre(
  c: Coffre,
  claim: Claim,
  federation: Parameters<typeof verifierTeteReseau>[1],
): Reclamation {
  const deja = new Set(c.tour.coffres ?? []);
  const v = jugerClaim(claim, federation, deja);
  if (!v.ok) return { ok: false, motif: v.motif };
  const tire = coffreDe(v.idBloc, claim.piece, claim.tete.hauteur);
  return {
    ok: true,
    coffre: {
      ...c,
      objets: [...c.objets, ...tire.objets],
      tour: { ...c.tour, coffres: [...deja, v.cle] },
    },
    ouverture: {
      cle: v.cle,
      idBloc: v.idBloc,
      hauteur: claim.tete.hauteur,
      graine: tire.graine,
      tier: tire.tier,
      objets: tire.objets,
    },
  };
}

File: labo/coffre_horaire.py
#!/usr/bin/env python3
"""Coffre horaire — un coffre par bloc, réclamé avec une pièce (docs/SPEC_COFFRE_HORAIRE.md).
Figures, pas preuves : rien n'est écrit sur la chaîne ; un coffre se juge hors ligne comme une
ascension. Bibliothèque standard. Rejoue le tirage sur les VRAIES têtes du testnet.

    graine  = sha256d("eidos-coffre/1" ‖ id_bloc(32) ‖ txid(32) ‖ rang(4, gros-boutiste))
    tier    = 1 + zéros de tête du premier octet de la graine       ∈ 1..9
              P(tier = t) = 2^−t pour t = 1..8, P(tier = 9) = 2^−8   (somme = 1, exactement)
    objets  = tier objets ; l'objet j part de h_j = sha256d(graine ‖ j(1)) et d'un âge par tier
              (1–3 Kali, 4–5 Dvâpara, 6–7 Trétâ, 8–9 Satya). Le labo s'arrête là : un objet est
              un ObjetPorte (mot, archétype, genre, emplacement), dérivé côté atelier par
              objetDepuisGraine + habille, comme un tirage de bloc. Le labo n'a pas de mots.
    règle   = une pièce, un bloc, un coffre ; le sac a 27 places, le surplus est perdu.

Usage : python3 labo/coffre_horaire.py           (rejoue chaine-eidos.dat × etat.json.sorties)
        python3 labo/coffre_horaire.py --vecteurs  (écrit labo/coffre_vecteurs.json pour le port TS)
MESURE DE LA RAFALE (spec §5) : `rafale()` compte ce qu'une pièce neuve récolte en réclamant
d'un coup tous les blocs passés, et le compare à une fenêtre d'un jour (24 blocs). C'est cette
mesure qui décide d'adopter la fenêtre ou non — voir la conclusion imprimée par ce module.

LIMITE : le juge hors ligne d'un claim (tête XMSS + pièce Merkle contre utxo_root_h) est celui de
l'ascension (ancrage.ts) ; ici on ne rejoue que le tirage. Une pièce peut réclamer les blocs passés
d'un coup : seule la taille du sac borne cette rafale — limite assumée, voir spec §4."""
import hashlib, json, os, sys
if hasattr(sys.stdout, "reconfigure"): sys.stdout.reconfigure(encoding="utf-8")
ICI = os.path.dirname(os.path.abspath(__file__)); RACINE = os.path.dirname(ICI)
sys.path.insert(0, RACINE); sys.path.insert(0, ICI)

TAG = b"eidos-coffre/1"
AGES = ("Kali", "Kali", "Kali", "Dvapara", "Dvapara", "Treta", "Treta", "Satya", "Satya")
SAC_PLACES, TIERS = 27, 9
PROBA = [2 ** -t for t in range(1, 9)] + [2 ** -8]

def sha256d(b): return hashlib.sha256(hashlib.sha256(b).digest()).digest()

def graine_coffre(id_bloc_hex, txid_hex, rang):
    return sha256d(TAG + bytes.fromhex(id_bloc_hex) + bytes.fromhex(txid_hex) + int(rang).to_bytes(4, "big"))

def tier_de(graine):
    b = graine[0]
    return 1 + (8 if b == 0 else 8 - b.bit_length())

def coffre(id_bloc_hex, txid_hex, rang):
    g = graine_coffre(id_bloc_hex, txid_hex, rang); t = tier_de(g)
    # une graine par objet ; l'objet lui-meme est derive cote atelier (voir en-tete)
    objets = [{"graine": sha256d(g + bytes([j])).hex(), "age": AGES[t - 1]} for j in range(t)]
    return {"graine": g.hex(), "tier": t, "objets": objets}

class Rejet(ValueError): pass

def reclamer(sac, deja, id_bloc_hex, txid_hex, rang):
    """Un claim : refuse la seconde fois pour la même (pièce, bloc) ; remplit le sac, perd le surplus."""
    cle = (txid_hex, int(rang), id_bloc_hex)
    if cle in deja: raise Rejet(f"coffre déjà réclamé pour {txid_hex[:8]}:{rang} au bloc {id_bloc_hex[:8]}")
    deja.add(cle)
    c = coffre(id_bloc_hex, txid_hex, rang)
    place = SAC_PLACES - len(sac); perdus = max(0, len(c["objets"]) - place)
    sac.extend(c["objets"][:place])
    return c, perdus

# ---------- les têtes et les pièces réelles ----------
def tetes_reelles():
    import noeud as N, federation as F
    c, fed = N.config(); buf = open(os.path.join(RACINE, "chaine-eidos.dat"), "rb").read()
    ch = F.ChaineFederee(fed); i = 10; out = []
    while i < len(buf):
        blk, i = N.deser_bloc(buf, i)
        out.append({"hauteur": blk["height"], "id_bloc": ch.id_bloc(blk).hex(), "ts": blk["ts"]})
    return out

def pieces_reelles():
    e = json.load(open(os.path.join(RACINE, "etat.json"), encoding="utf-8"))
    return [{"txid": k.split(":")[0], "rang": int(k.split(":")[1]), **v} for k, v in e["sorties"].items()]

FENETRE_JOUR = 24   # blocs : un coffre de hauteur h ne se réclamerait qu'avec une tête ≤ h + 23

def rafale(tetes, txid, rang=0, fenetre=None):
    """Ce qu'une pièce récolte en réclamant les blocs passés d'un coup. `fenetre` = None : tout
    l'historique ; sinon les `fenetre` derniers blocs. Le sac (27) borne toujours la prise."""
    choisies = tetes if fenetre is None else tetes[-fenetre:]
    tires = [coffre(t["id_bloc"], txid, rang)["tier"] for t in choisies]
    offerts = sum(tires)
    return {"blocs": len(choisies), "offerts": offerts, "pris": min(offerts, SAC_PLACES),
            "perdus": max(0, offerts - SAC_PLACES), "meilleur": max(tires) if tires else 0}

# ---------- LAB TEST ----------
def lab_test(tetes, pieces):
    R = {}; n = 0; comptes = [0] * TIERS; deja = set()
    for t in tetes:
        for p in pieces:
            c = coffre(t["id_bloc"], p["txid"], p["rang"]); comptes[c["tier"] - 1] += 1; n += 1
    R["K39_tier_borne"] = all(1 <= tier_de(bytes([b]) + b"\0" * 31) <= 9 for b in range(256)) and \
        [tier_de(bytes([b]) + b"\0" * 31) for b in (255, 128, 127, 64, 1, 0)] == [1, 1, 2, 2, 8, 9]
    R["K40_probas_somment_a_1"] = abs(sum(PROBA) - 1) < 1e-15 and \
        all(abs(sum(1 for b in range(256) if tier_de(bytes([b]) + b"\0" * 31) == t) / 256 - PROBA[t - 1]) < 1e-15 for t in range(1, 10))
    # sur les claims réels : chaque tier observé dans ±4σ de l'attendu (binomiale), tiers 1..3 seulement (les autres trop rares)
    def ok(t): 
        mu = n * PROBA[t - 1]; sig = (mu * (1 - PROBA[t - 1])) ** 0.5
        return abs(comptes[t - 1] - mu) <= 4 * sig + 1
    R["K41_distribution_reelle_plausible"] = n >= 100 and all(ok(t) for t in (1, 2, 3))
    R["K42_deterministe"] = coffre(tetes[0]["id_bloc"], pieces[0]["txid"], pieces[0]["rang"]) == \
        coffre(tetes[0]["id_bloc"], pieces[0]["txid"], pieces[0]["rang"])
    sac = []; deja = set()
    c, _ = reclamer(sac, deja, tetes[0]["id_bloc"], pieces[0]["txid"], pieces[0]["rang"])
    try: reclamer(sac, deja, tetes[0]["id_bloc"], pieces[0]["txid"], pieces[0]["rang"]); R["K43_une_piece_un_bloc"] = False
    except Rejet: R["K43_une_piece_un_bloc"] = True
    sac = []; deja = set(); perdus = 0
    for t in tetes: perdus += reclamer(sac, deja, t["id_bloc"], pieces[0]["txid"], pieces[0]["rang"])[1]
    R["K44_sac_borne_27"] = len(sac) <= SAC_PLACES and perdus == max(0, sum(coffre(t["id_bloc"], pieces[0]["txid"], pieces[0]["rang"])["tier"] for t in tetes) - SAC_PLACES)
    R["K45_objets_egal_tier"] = all(len(coffre(t["id_bloc"], p["txid"], p["rang"])["objets"]) == coffre(t["id_bloc"], p["txid"], p["rang"])["tier"] for t in tetes[:3] for p in pieces[:3])
    # K47 — la rafale rétroactive : mesurée, pas supposée. Le sac plafonne la prise à 27 dans les
    # deux cas dès que l'historique dépasse ~14 blocs ; la fenêtre d'un jour ne change donc RIEN
    # à ce qu'une pièce neuve emporte. Elle n'est pas adoptée (spec §5).
    r_tout = [rafale(tetes, p["txid"], p["rang"]) for p in pieces]
    r_jour = [rafale(tetes, p["txid"], p["rang"], FENETRE_JOUR) for p in pieces]
    R["K47_fenetre_sans_effet_sur_la_prise"] = len(tetes) > FENETRE_JOUR and \
        all(a["pris"] == b["pris"] == SAC_PLACES for a, b in zip(r_tout, r_jour))
    R["K46_tetes_reelles_heure_par_heure"] = len(tetes) >= 2 and all(b["ts"] - a["ts"] >= 3600 and b["hauteur"] == a["hauteur"] + 1 for a, b in zip(tetes, tetes[1:]))
    return R, n, comptes

if __name__ == "__main__":
    tetes, pieces = tetes_reelles(), pieces_reelles()
    if "--vecteurs" in sys.argv:
        vec = [{"id_bloc": t["id_bloc"], "txid": p["txid"], "rang": p["rang"], **coffre(t["id_bloc"], p["txid"], p["rang"])}
               for t in tetes[:3] for p in pieces[:3]]
        json.dump(vec, open(os.path.join(ICI, "coffre_vecteurs.json"), "w", encoding="utf-8", newline="\n"), indent=0)
        print("labo/coffre_vecteurs.json :", len(vec), "vecteurs"); sys.exit(0)
    R, n, comptes = lab_test(tetes, pieces)
    for k, ok in R.items(): print(("PASS " if ok else "FAIL "), k)
    print(f"{len(tetes)} têtes réelles × {len(pieces)} pièces = {n} coffres ; tiers 1..9 :", comptes,
          "| attendus :", [round(n * p, 1) for p in PROBA])
    a = rafale(tetes, pieces[0]["txid"], pieces[0]["rang"])
    b = rafale(tetes, pieces[0]["txid"], pieces[0]["rang"], FENETRE_JOUR)
    print(f"rafale d'une pièce neuve — tout l'historique ({a['blocs']} blocs) : {a['offerts']} objets offerts, "
          f"{a['pris']} pris, {a['perdus']} perdus ; fenêtre d'un jour ({b['blocs']} blocs) : "
          f"{b['offerts']} offerts, {b['pris']} pris. Le sac plafonne les deux : la fenêtre ne change rien.")
    sys.exit(0 if all(R.values()) else 1)
